### ISSUE TITLE
Error: Cannot find a frame to evict from

### ISSUE THREAD

**Issue by uwegeercken** [kuzudb/kuzu#1374]
This query works ok in the command client:
```
kuzu> match(c:Country)<-[r2:isPartOf]-(g:Geoname)-[r:isTypeOf]->(f:Feature) 
..> where f.code='AIRP' 
..> with c, count(r) as numberOfAirports 
..> where numberOfAirports >100 
..> return c.name as country, numberOfAirports;
```

But when I add the Node "Continent" and want to aggregate the number of Airports by Continent, I get an error:

```
kuzu> match(co:Continent)<-[r3:inContinent]-(c:Country)<-[r2:isPartOf]-(g:Geoname)-[r:isTypeOf]->(f:Feature) 
..> where f.code='AIRP'
..> with co,c,count(r) as numberOfAirports
..> return co.name, c.code, numberOfAirports;
Error: Buffer manager exception: Cannot find a frame to evict from.
```

Can somebody explain please, what this error message means? 

Is it about buffer manager size? And if yes what is the default setting for this value (in bytes)?

Greetings

**Comment by ray6080** [comment]
Hi @uwegeercken , thanks for reporting! Yes it is about buffer manager (bm) size. by default, we set bm size to 0.8 * physical available memory.  When you start the shell, you can manually specify bm size in MB with the option `-d`. Can you try increase it first? Also, if that doesn't work, you can also try to lower the num of threads for execution (inside kuzu shell, `:thread` can limit num of threads for execution).

Let me know if you cannot fix the problem by tuning the bm and num of threads.

**Comment by andyfengHKU** [comment]
This is most likely because we pick a plan which requires excessive amount of memory. It becomes harder to pick an optimal plan as number of joins (i.e. graph pattern) increases.

Can u do `EXPLAIN match ...` and show me the plan that is being picked? And the number of rows of each table (node and rely) involved in this query if possible.

TBH I don't have a quick work around to pick a different plan in release v0.0.2. In your case changing the return clause might help 
```
with co.ID, co.name, c.ID, c.code, count(r) as numberOfAirports
return co.name, c.code, numberOfAirports
```
because `with co, c, ..` will try materializing all columns of the two tables. 

If this still doesn't work, I'll need know your table statistics and check if something goes wrong when analyzing the cost of plan.

**Comment by uwegeercken** [comment]
@ray6080 Thanks for the reply. 

- I think 0.8 of physical available memory is already a lot. How can I increase that without putting my computer into a state where it does not respond anymore  (happened already to me)?
- it would be good to document the defualts for both memory and threads and also have the client help page display it.
- why is the -d startup option in megabytes and the :buffer_manager_size option in bytes?

**Comment by uwegeercken** [comment]
@andyfengHKU  

Here are the total counts for all nodes and rels involved:
```
kuzu> match(n:Continent) return count(n) as total_nodes;
---------------
| total_nodes |
---------------
| 7           |
---------------

kuzu> match()-[r:inContinent]->() return count(r) as total_rels;
--------------
| total_rels |
--------------
| 193        |
--------------

kuzu> match(n:Country) return count(n) as total_nodes;
---------------
| total_nodes |
---------------
| 193         |
---------------

kuzu> match()-[r:isPartOf]->() return count(r) as total_rels;
--------------
| total_rels |
--------------
| 11815128   |
--------------

kuzu> match(n:Geoname) return count(n) as total_nodes;
---------------
| total_nodes |
---------------
| 11923604    |
---------------

kuzu> match()-[r:isTypeOf]->() return count(r) as total_rels;
--------------
| total_rels |
--------------
| 11828537   |
--------------

kuzu> match(n:Feature) return count(n) as total_nodes;
---------------
| total_nodes |
---------------
| 680         |
---------------

```

**Comment by uwegeercken** [comment]
@andyfengHKU executing the explain match gives me this:
```
┌─────────────────────────────────┐
│┌───────────────────────────────┐│
││         Physical Plan         ││
│└───────────────────────────────┘│
└─────────────────────────────────┘
┌─────────────────────────────────┐                                                                                                                                                                                    
│        RESULT_COLLECTOR         │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│             co.name             │                                                                                                                                                                                    
│             c.code              │                                                                                                                                                                                    
│        numberOfAirports         │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       NumOutputTuples: 0        │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│     ExecutionTime: 0.000000     │                                                                                                                                                                                    
└────────────────┬────────────────┘                                                                                                                                                                                    
┌────────────────┴────────────────┐                                                                                                                                                                                    
│           PROJECTION            │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│             co.name             │                                                                                                                                                                                    
│             c.code              │                                                                                                                                                                                    
│        numberOfAirports         │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       NumOutputTuples: 0        │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│     ExecutionTime: 0.000000     │                                                                                                                                                                                    
└────────────────┬────────────────┘                                                                                                                                                                                    
┌────────────────┴────────────────┐                                                                                                                                                                                    
│           PROJECTION            │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│             co._id              │                                                                                                                                                                                    
│             co.name             │                                                                                                                                                                                    
│              c._id              │                                                                                                                                                                                    
│             c.code              │                                                                                                                                                                                    
│        numberOfAirports         │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       NumOutputTuples: 0        │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│     ExecutionTime: 0.000000     │                                                                                                                                                                                    
└────────────────┬────────────────┘                                                                                                                                                                                    
┌────────────────┴────────────────┐                                                                                                                                                                                    
│         AGGREGATE_SCAN          │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       Group By [_0_co._id       │                                                                                                                                                                                    
│            _0_co.name           │                                                                                                                                                                                    
│             _1_c._id            │                                                                                                                                                                                    
│            _1_c.code            │                                                                                                                                                                                    
│                ]                │                                                                                                                                                                                    
│    Aggregate [COUNT(_6_r._id)   │                                                                                                                                                                                    
│                ]                │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       NumOutputTuples: 0        │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│     ExecutionTime: 0.000000     │                                                                                                                                                                                    
└────────────────┬────────────────┘                                                                                                                                                                                    
┌────────────────┴────────────────┐                                                                                                                                                                                    
│            AGGREGATE            │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       Group By [_0_co._id       │                                                                                                                                                                                    
│            _0_co.name           │                                                                                                                                                                                    
│             _1_c._id            │                                                                                                                                                                                    
│            _1_c.code            │                                                                                                                                                                                    
│                ]                │                                                                                                                                                                                    
│    Aggregate [COUNT(_6_r._id)   │                                                                                                                                                                                    
│                ]                │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       NumOutputTuples: 0        │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│     ExecutionTime: 0.000000     │                                                                                                                                                                                    
└────────────────┬────────────────┘                                                                                                                                                                                    
┌────────────────┴────────────────┐                                                                                                                                                                                    
│           PROJECTION            │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│              r._id              │                                                                                                                                                                                    
│             co._id              │                                                                                                                                                                                    
│             co.name             │                                                                                                                                                                                    
│              c._id              │                                                                                                                                                                                    
│             c.code              │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│       NumOutputTuples: 0        │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│     ExecutionTime: 0.000000     │                                                                                                                                                                                    
└────────────────┬────────────────┘                                                                                                                                                                                    
┌────────────────┴────────────────┐                                                                                                                                                                                    
│         HASH_JOIN_PROBE         │                                                                                                                                                                                    
│   ---------------------------   │                                                                                                                                                                                    
│              f._id              │                                                                                                                                                                                    
│   ---------------------------   │──────────────────┐                                                                                                                                                                 
│       NumOutputTuples: 0        │                  │                                                                                                                                                                 
│   ---------------------------   │                  │                                                                                                                                                                 
│     ExecutionTime: 0.000000     │                  │                                                                                                                                                                 
└────────────────┬────────────────┘                  │                                                                                                                                                                 
┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                                                                                                                                
│             FLATTEN             │ │         HASH_JOIN_BUILD         │                                                                                                                                                
│   ---------------------------   │ │   ---------------------------   │                                                                                                                                                
│            _5_f._id             │ │              f._id              │                                                                                                                                                
│   ---------------------------   │ │   ---------------------------   │                                                                                                                                                
│       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                                                                                                                                
│   ---------------------------   │ │   ---------------------------   │                                                                                                                                                
│     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                                                                                                                                
└────────────────┬────────────────┘ └────────────────┬────────────────┘                                                                                                                                                
┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                                                                                                                                
│             FILTER              │ │         HASH_JOIN_PROBE         │                                                                                                                                                
│   ---------------------------   │ │   ---------------------------   │                                                                                                                                                
│          f.code='AIRP'          │ │              g._id              │                                                                                                                                                
│   ---------------------------   │ │   ---------------------------   │                                                                        ──────────────────────────────────────────────────────┐                 
│       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                                                                                                              │                 
│   ---------------------------   │ │   ---------------------------   │                                                                                                                              │                 
│     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                                                                                                              │                 
└────────────────┬────────────────┘ └────────────────┬────────────────┘                                                                                                                              │                 
┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                                                                                             ┌────────────────┴────────────────┐
│       SCAN_NODE_PROPERTY        │ │             FLATTEN             │                                                                                                             │         HASH_JOIN_BUILD         │
│   ---------------------------   │ │   ---------------------------   │                                                                                                             │   ---------------------------   │
│             f.code              │ │            _3_g._id             │                                                                                                             │              g._id              │
│   ---------------------------   │ │   ---------------------------   │                                                                                                             │   ---------------------------   │
│       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                                                                                             │       NumOutputTuples: 0        │
│   ---------------------------   │ │   ---------------------------   │                                                                                                             │   ---------------------------   │
│     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                                                                                             │     ExecutionTime: 0.000000     │
└────────────────┬────────────────┘ └────────────────┬────────────────┘                                                                                                             └────────────────┬────────────────┘
┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                                                                                             ┌────────────────┴────────────────┐
│          SCAN_NODE_ID           │ │         HASH_JOIN_PROBE         │                                                                                                             │      SCAN_REL_TABLE_LISTS       │
│   ---------------------------   │ │   ---------------------------   │                                                                                                             │   ---------------------------   │
│                f                │ │             co._id              │                                                                                                             │              f<-g               │
│   ---------------------------   │ │   ---------------------------   │──────────────────┐                                                                                          │   ---------------------------   │
│       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                  │                                                                                          │       NumOutputTuples: 0        │
│   ---------------------------   │ │   ---------------------------   │                  │                                                                                          │   ---------------------------   │
│     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                  │                                                                                          │     ExecutionTime: 0.000000     │
└─────────────────────────────────┘ └────────────────┬────────────────┘                  │                                                                                          └────────────────┬────────────────┘
                                    ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                                                         ┌────────────────┴────────────────┐
                                    │             FLATTEN             │ │         HASH_JOIN_BUILD         │                                                                         │             FLATTEN             │
                                    │   ---------------------------   │ │   ---------------------------   │                                                                         │   ---------------------------   │
                                    │            _0_co._id            │ │             co._id              │                                                                         │            _5_f._id             │
                                    │   ---------------------------   │ │   ---------------------------   │                                                                         │   ---------------------------   │
                                    │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                                                         │       NumOutputTuples: 0        │
                                    │   ---------------------------   │ │   ---------------------------   │                                                                         │   ---------------------------   │
                                    │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                                                         │     ExecutionTime: 0.000000     │
                                    └────────────────┬────────────────┘ └────────────────┬────────────────┘                                                                         └────────────────┬────────────────┘
                                    ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                                                         ┌────────────────┴────────────────┐
                                    │       SCAN_NODE_PROPERTY        │ │         HASH_JOIN_PROBE         │                                                                         │          SCAN_NODE_ID           │
                                    │   ---------------------------   │ │   ---------------------------   │                                                                         │   ---------------------------   │
                                    │             co.name             │ │              c._id              │                                                                         │                f                │
                                    │   ---------------------------   │ │   ---------------------------   │──────────────────────────────────────────────────────┐                  │   ---------------------------   │
                                    │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                                      │                  │       NumOutputTuples: 0        │
                                    │   ---------------------------   │ │   ---------------------------   │                                                      │                  │   ---------------------------   │
                                    │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                                      │                  │     ExecutionTime: 0.000000     │
                                    └────────────────┬────────────────┘ └────────────────┬────────────────┘                                                      │                  └─────────────────────────────────┘
                                    ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                     ┌────────────────┴────────────────┐                                    
                                    │          SCAN_NODE_ID           │ │         HASH_JOIN_PROBE         │                                     │         HASH_JOIN_BUILD         │                                    
                                    │   ---------------------------   │ │   ---------------------------   │                                     │   ---------------------------   │                                    
                                    │               co                │ │              c._id              │                                     │              c._id              │                                    
                                    │   ---------------------------   │ │   ---------------------------   │──────────────────┐                  │   ---------------------------   │                                    
                                    │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                  │                  │       NumOutputTuples: 0        │                                    
                                    │   ---------------------------   │ │   ---------------------------   │                  │                  │   ---------------------------   │                                    
                                    │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                  │                  │     ExecutionTime: 0.000000     │                                    
                                    └─────────────────────────────────┘ └────────────────┬────────────────┘                  │                  └────────────────┬────────────────┘                                    
                                                                        ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                    
                                                                        │             FLATTEN             │ │         HASH_JOIN_BUILD         │ │      SCAN_REL_TABLE_LISTS       │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │            _1_c._id             │ │              c._id              │ │              co<-c              │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                    
                                                                        └────────────────┬────────────────┘ └────────────────┬────────────────┘ └────────────────┬────────────────┘                                    
                                                                        ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                    
                                                                        │       SCAN_NODE_PROPERTY        │ │      SCAN_REL_TABLE_LISTS       │ │             FLATTEN             │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │             c.code              │ │              g->c               │ │            _0_co._id            │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                    
                                                                        └────────────────┬────────────────┘ └────────────────┬────────────────┘ └────────────────┬────────────────┘                                    
                                                                        ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐ ┌────────────────┴────────────────┐                                    
                                                                        │          SCAN_NODE_ID           │ │             FLATTEN             │ │          SCAN_NODE_ID           │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │                c                │ │            _3_g._id             │ │               co                │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │ │       NumOutputTuples: 0        │                                    
                                                                        │   ---------------------------   │ │   ---------------------------   │ │   ---------------------------   │                                    
                                                                        │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │ │     ExecutionTime: 0.000000     │                                    
                                                                        └─────────────────────────────────┘ └────────────────┬────────────────┘ └─────────────────────────────────┘                                    
                                                                                                            ┌────────────────┴────────────────┐                                                                        
                                                                                                            │          SCAN_NODE_ID           │                                                                        
                                                                                                            │   ---------------------------   │                                                                        
                                                                                                            │                g                │                                                                        
                                                                                                            │   ---------------------------   │                                                                        
                                                                                                            │       NumOutputTuples: 0        │                                                                        
                                                                                                            │   ---------------------------   │                                                                        
                                                                                                            │     ExecutionTime: 0.000000     │                                                                        
                                                                                                            └─────────────────────────────────┘            
```

**Comment by uwegeercken** [comment]
@andyfengHKU but actually the query is not good, as I want the result per continent and not by continent and country. So I changed the query:
```
match(co:Continent)<-[r3:inContinent]-(c:Country)<-[r2:isPartOf]-(g:Geoname)-[r:isTypeOf]->(f:Feature) 
where f.code='AIRP'
with co,c,count(r) as numberOfAirports
return co.name, numberOfAirports;
```

to:

```
match(co:Continent)<-[r3:inContinent]-(c:Country)<-[r2:isPartOf]-(g:Geoname)-[r:isTypeOf]->(f:Feature)  
where f.code='AIRP'
with co,c,count(r) as numberOfAirports
return co.name, numberOfAirports;
```
but the query runs forever. So I have changed the query query like this:
```
match(c:Country)<-[r2:isPartOf]-(g:Geoname)-[r:isTypeOf]->(f:Feature)  
where f.code='AIRP'
with c,r
match (co:Continent)<-[r3:inContinent]-(c)
return co.name, count(r) as numberOfAirports;

```
Here I match the result of the first query and then do the aggragate. In this case the query returns the result in about 2 secs:
```
kuzu> match(c:Country)<-[r2:isPartOf]-(g:Geoname)-[r:isTypeOf]->(f:Feature)   
..> where f.code='AIRP'
..> with c,r
..> match (co:Continent)<-[r3:inContinent]-(c)
..> return co.name, count(r) as numberOfAirports;
------------------------------------
| co.name       | numberOfAirports |
------------------------------------
| Africa        | 1217             |
------------------------------------
| Asia          | 1862             |
------------------------------------
| Europe        | 1808             |
------------------------------------
| South America | 1225             |
------------------------------------
| North America | 15937            |
------------------------------------
| Oceania       | 1161             |
------------------------------------

```
I hope this makes some sense to you. Otherwise let me know if I should provide more details.

**Comment by andyfengHKU** [comment]
Hi @uwegeercken, you are right. We end up picking a terrible plan for query `(co:Continent)<-[r3:inContinent]-(c:Country)<-[r2:isPartOf]-(g:Geoname)-[r:isTypeOf]->(f:Feature)`. I'll add this case in our test suite and try fixing it in our next release.

**Comment by ray6080** [comment]
> @ray6080 Thanks for the reply.
> 
> * I think 0.8 of physical available memory is already a lot. How can I increase that without putting my computer into a state where it does not respond anymore  (happened already to me)?
> * it would be good to document the defualts for both memory and threads and also have the client help page display it.
> * why is the -d startup option in megabytes and the :buffer_manager_size option in bytes?

Hi @uwegeercken ,sorry for the late reply. You are right, increasing bm size doesn't help. For the second suggestion, yes, that's a good idea, thanks for the suggestion. And for the third comment, we actually don't think resizing bm inside shell is a good idea, and removed that feature in recent master. Thanks!

**Comment by andyfengHKU** [comment]
Hi @uwegeercken, this should be fixed in our release v0.0.3
